File: llm_written_v1/claude-sonnet-4.5/smith-waterman/c++/test_smith_waterman.cpp
#include "smith_waterman.h"
#include <iostream>
#include <iomanip>
#include <string>
#include <vector>
#include <tuple>
#include <cmath>
#include <chrono>
#include <functional>

// Hash function for alignment results
size_t hashAlignment(const std::string& a1, const std::string& a2, int score, double pct) {
    std::hash<std::string> hasher;
    size_t h = hasher(a1);
    h ^= hasher(a2) + 0x9e3779b9 + (h << 6) + (h >> 2);
    h ^= std::hash<int>{}(score) + 0x9e3779b9 + (h << 6) + (h >> 2);
    h ^= std::hash<double>{}(pct) + 0x9e3779b9 + (h << 6) + (h >> 2);
    return h;
}

struct TestCase {
    std::string name;
    std::string query;
    std::string reference;
};

std::vector<TestCase> getTestCases() {
    return {
        // Edge case: empty strings
        {"edge_empty", "", ""},
        
        // Edge case: one empty
        {"edge_one_empty", "ACGT", ""},
        
        // Edge case: single character
        {"edge_single", "A", "A"},
        
        // Small: exact match
        {"small_exact", "ACGT", "ACGT"},
        
        // Small: no match
        {"small_nomatch", "AAAA", "TTTT"},
        
        // Small: partial match
        {"small_partial", "ACGTACGT", "ACTTACGT"},
        
        // Medium: realistic sequences ~100 bp
        {"medium_dna", 
         "GCTAGCTACGATCGATCGATCGATCGATCGATCGTAGCTAGCTAGCTAGCTAGCTAGCTAGCTAGCTAGCTAGCTAGCTAGCTAGCTAGCTAGCT",
         "GCTAGCTACGATCGATCGATCGATCGATCGATCGTAGCTAGCTAGCTAGCTAGCTAGCTAGCTAGCTAGCTAGCTAGCTAGCTAGCTAGCTAGCT"},
        
        {"medium_mutation",
         "ATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGAT",
         "ATCGATCGATCGATCGATCGATTTTTTTCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGAT"},
        
        // Large: ~1000 bp
        {"large_seq",
         "ATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGAT"
         "ATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGAT"
         "ATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGAT"
         "ATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGAT"
         "ATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGAT"
         "ATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGAT"
         "ATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGAT"
         "ATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGAT"
         "ATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGAT"
         "ATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGAT",
         "ATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGAT"
         "ATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGAT"
         "ATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGAT"
         "ATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGAT"
         "TTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTT"
         "ATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGAT"
         "ATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGAT"
         "ATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGAT"
         "ATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGAT"
         "ATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGAT"}
    };
}

int main(int argc, char* argv[]) {
    bool runPerf = (argc > 1 && std::string(argv[1]) == "--perf");
    
    std::cout << "=== Smith-Waterman Differential Test Suite ===" << std::endl;
    std::cout << "Testing correctness, determinism";
    if (runPerf) std::cout << ", and performance";
    std::cout << std::endl << std::endl;
    
    SmithWaterman sw(2, -1, -1);
    auto testCases = getTestCases();
    
    bool allPassed = true;
    int passCount = 0;
    int totalTests = testCases.size();
    
    for (const auto& tc : testCases) {
        std::cout << "Test: " << tc.name << " (query=" << tc.query.length() 
                  << ", ref=" << tc.reference.length() << ")" << std::endl;
        
        // Run once
        auto [a1_r1, a2_r1, score_r1, pct_r1] = sw.findAlignment(tc.query, tc.reference);
        size_t hash1 = hashAlignment(a1_r1, a2_r1, score_r1, pct_r1);
        
        // Run again for determinism
        auto [a1_r2, a2_r2, score_r2, pct_r2] = sw.findAlignment(tc.query, tc.reference);
        size_t hash2 = hashAlignment(a1_r2, a2_r2, score_r2, pct_r2);
        
        bool deterministicPass = (hash1 == hash2);
        
        std::cout << "  Result: score=" << score_r1 << ", pct=" << std::fixed 
                  << std::setprecision(2) << pct_r1 << "%" << std::endl;
        std::cout << "  Hash run1: " << hash1 << std::endl;
        std::cout << "  Hash run2: " << hash2 << std::endl;
        std::cout << "  Determinism: " << (deterministicPass ? "✓ PASS" : "✗ FAIL") << std::endl;
        
        if (deterministicPass) {
            passCount++;
        } else {
            allPassed = false;
        }
        
        // Performance test for large case
        if (runPerf && tc.name == "large_seq") {
            std::cout << "  Performance measurement:" << std::endl;
            
            auto start = std::chrono::high_resolution_clock::now();
            auto result = sw.findAlignment(tc.query, tc.reference);
            auto end = std::chrono::high_resolution_clock::now();
            
            double duration = std::chrono::duration<double>(end - start).count();
            std::cout << "    Time: " << std::fixed << std::setprecision(4) 
                      << duration << " sec" << std::endl;
        }
        
        std::cout << std::endl;
    }
    
    std::cout << "=== Summary ===" << std::endl;
    std::cout << "Tests passed: " << passCount << "/" << totalTests << std::endl;
    
    if (allPassed) {
        std::cout << "✓ All tests PASSED" << std::endl;
        return 0;
    } else {
        std::cout << "✗ Some tests FAILED" << std::endl;
        return 1;
    }
}

File: llm_written_v1/claude-sonnet-4.5/smith-waterman/c#/test_smithwaterman.cs
using System;
using System.Diagnostics;
using System.Linq;
using System.Security.Cryptography;
using System.Text;
using SmithWatermanAlignment;

class TestSmithWaterman
{
    static int exitCode = 0;

    static void Main(string[] args)
    {
        Console.WriteLine("=== Smith-Waterman Differential Test Suite ===\n");

        // Test cases: edge, small, medium, large
        RunTestCase("Edge: Empty strings", "", "", 2, -1, -1);
        RunTestCase("Edge: Single char match", "A", "A", 2, -1, -1);
        RunTestCase("Edge: Single char mismatch", "A", "C", 2, -1, -1);
        RunTestCase("Edge: Query empty", "", "ACGT", 2, -1, -1);
        RunTestCase("Edge: Reference empty", "ACGT", "", 2, -1, -1);

        RunTestCase("Small: Short exact match", "ACGT", "ACGT", 2, -1, -1);
        RunTestCase("Small: With mismatch", "ACGT", "AGGT", 2, -1, -1);
        RunTestCase("Small: With gap", "ACGT", "ACT", 2, -1, -1);
        RunTestCase("Small: No similarity", "AAAA", "CCCC", 2, -1, -1);

        // Medium test case
        string query1 = GenerateRandomDNA(200, 42);
        string ref1 = GenerateRandomDNA(200, 43);
        RunTestCase("Medium: Random 200x200", query1, ref1, 2, -1, -1);

        // Large test case (below threshold - will use sequential)
        string query2 = GenerateRandomDNA(400, 100);
        string ref2 = GenerateRandomDNA(400, 101);
        RunTestCase("Large: Random 400x400", query2, ref2, 2, -1, -1);

        // XLarge test case (above threshold - will use parallel)
        string query3 = GenerateRandomDNA(600, 200);
        string ref3 = GenerateRandomDNA(600, 201);
        RunTestCase("XLarge: Random 600x600", query3, ref3, 2, -1, -1, true);

        // XXLarge for performance measurement
        string query4 = GenerateRandomDNA(1200, 300);
        string ref4 = GenerateRandomDNA(1200, 301);
        RunTestCase("XXLarge: Random 1200x1200", query4, ref4, 2, -1, -1, true);

        Console.WriteLine("\n=== Test Summary ===");
        if (exitCode == 0)
        {
            Console.WriteLine("✓ All tests PASSED");
        }
        else
        {
            Console.WriteLine($"✗ Tests FAILED (exit code: {exitCode})");
        }

        Environment.Exit(exitCode);
    }

    static void RunTestCase(string name, string query, string reference, 
                           int match, int mismatch, int gap, bool measurePerf = false)
    {
        Console.WriteLine($"Test: {name}");
        Console.WriteLine($"  Query length: {query.Length}, Reference length: {reference.Length}");

        // Run sequential baseline
        var seqAlgo = new SmithWatermanSequential(match, mismatch, gap);
        Stopwatch swSeq = Stopwatch.StartNew();
        var seqResult = seqAlgo.FindAlignment(query, reference);
        swSeq.Stop();
        double seqTime = swSeq.Elapsed.TotalMilliseconds;

        // Run parallel version twice for determinism check
        var parAlgo = new SmithWaterman(match, mismatch, gap);
        
        Stopwatch swPar1 = Stopwatch.StartNew();
        var parResult1 = parAlgo.FindAlignment(query, reference);
        swPar1.Stop();
        double parTime1 = swPar1.Elapsed.TotalMilliseconds;

        Stopwatch swPar2 = Stopwatch.StartNew();
        var parResult2 = parAlgo.FindAlignment(query, reference);
        swPar2.Stop();
        double parTime2 = swPar2.Elapsed.TotalMilliseconds;

        // Check correctness (parallel vs sequential)
        bool correctnessPass = CompareResults(seqResult, parResult1, "Sequential", "Parallel-Run1");

        // Check determinism (parallel run1 vs run2)
        bool determinismPass = CompareResults(parResult1, parResult2, "Parallel-Run1", "Parallel-Run2");

        if (correctnessPass && determinismPass)
        {
            Console.WriteLine("  ✓ PASS - Correctness and Determinism verified");
            
            // Show hashes for determinism proof
            string hash1 = ComputeHash($"{parResult1.Item1}|{parResult1.Item2}|{parResult1.Item3}|{parResult1.Item4}");
            string hash2 = ComputeHash($"{parResult2.Item1}|{parResult2.Item2}|{parResult2.Item3}|{parResult2.Item4}");
            Console.WriteLine($"    Parallel-Run1 hash: {hash1}");
            Console.WriteLine($"    Parallel-Run2 hash: {hash2}");
        }
        else
        {
            Console.WriteLine("  ✗ FAIL");
            exitCode = 1;
        }

        if (measurePerf && query.Length >= 500)
        {
            double avgParTime = (parTime1 + parTime2) / 2.0;
            double speedup = seqTime / avgParTime;
            Console.WriteLine($"  Performance: Seq={seqTime:F2}ms, Par={avgParTime:F2}ms, Speedup={speedup:F2}x, Cores={Environment.ProcessorCount}");
        }

        Console.WriteLine();
    }

    static bool CompareResults((string, string, int, double) result1, 
                              (string, string, int, double) result2,
                              string label1, string label2)
    {
        bool match = result1.Item1 == result2.Item1 &&
                     result1.Item2 == result2.Item2 &&
                     result1.Item3 == result2.Item3 &&
                     Math.Abs(result1.Item4 - result2.Item4) < 1e-9;

        if (!match)
        {
            Console.WriteLine($"  ✗ Mismatch between {label1} and {label2}:");
            Console.WriteLine($"    {label1}: AlignedA='{result1.Item1}', AlignedB='{result1.Item2}', Score={result1.Item3}, Identity={result1.Item4:F2}%");
            Console.WriteLine($"    {label2}: AlignedA='{result2.Item1}', AlignedB='{result2.Item2}', Score={result2.Item3}, Identity={result2.Item4:F2}%");
            
            // Compute hashes for easier comparison
            string hash1 = ComputeHash($"{result1.Item1}|{result1.Item2}|{result1.Item3}|{result1.Item4}");
            string hash2 = ComputeHash($"{result2.Item1}|{result2.Item2}|{result2.Item3}|{result2.Item4}");
            Console.WriteLine($"    {label1} hash: {hash1}");
            Console.WriteLine($"    {label2} hash: {hash2}");
        }

        return match;
    }

    static string GenerateRandomDNA(int length, int seed)
    {
        Random rand = new Random(seed);
        char[] bases = { 'A', 'C', 'G', 'T' };
        StringBuilder sb = new StringBuilder(length);
        
        for (int i = 0; i < length; i++)
        {
            sb.Append(bases[rand.Next(bases.Length)]);
        }
        
        return sb.ToString();
    }

    static string ComputeHash(string input)
    {
        using (SHA256 sha256 = SHA256.Create())
        {
            byte[] bytes = sha256.ComputeHash(Encoding.UTF8.GetBytes(input));
            return BitConverter.ToString(bytes).Replace("-", "").Substring(0, 16);
        }
    }
}
